src/dftbp/dftb/mdftb.F90

```diff
@@ -0,0 +1,1575 @@
+    real(dp), allocatable :: f50AB(:,:,:,:,:,:,:)
+
+    !> add for the H and the E
+    real(dp), allocatable :: pot10x1Atom(:)
+    real(dp), allocatable :: pot20x2Atom(:)
+    real(dp), allocatable :: pot10x0Atom(:,:)
+    real(dp), allocatable :: pot11x1Atom(:,:)
+    real(dp), allocatable :: pot21x2Atom(:,:)
+    real(dp), allocatable :: pot20x0Atom(:,:,:)
+    real(dp), allocatable :: pot21x1Atom(:,:,:)
+    real(dp), allocatable :: pot22x2Atom(:,:,:)
+    !> add for the gradient
+    real(dp), allocatable :: pot30x0Atom(:,:,:,:)
+    real(dp), allocatable :: pot31x1Atom(:,:,:,:)
+    real(dp), allocatable :: pot32x2Atom(:,:,:,:)
+
+    !> total energy
+    real(dp) :: energyTT, energyMD, energyDD, energyMQ, energyDQ, energyQQ
+
+  contains
+    procedure :: updateCoords
+    procedure :: pullDeltaDQAtom
+    procedure :: pushDeltaDQAtom
+    procedure :: updateDeltaDQAtom
+    procedure :: updateDQPotentials
+    procedure :: addMultiExpanHamiltonian
+    procedure :: addMultiExpanEnergy
+    procedure :: addMultiExpanGradients
+    procedure :: addAtomicDipoleMoment
+    procedure :: addAtomicQuadrupoleMoment
+    procedure :: getMultiExpanInfo
+    procedure :: getOrbitalEquiv
+  end type TMdftb
+
+  !> Number of dipole components used in tblite library (x, y, z)
+  integer, parameter :: dimDipole = 3
+
+  !> Number of quadrupole components used in tblite library (xx, xy, yy, xz, yz, zz)
+  integer, parameter :: dimQuadrupole = 6
+
+contains
+
+  !> Get information on required multipolar contributions
+  subroutine getMultiExpanInfo(this, nDipole, nQuadrupole)
+
+    !> Data structure
+    class(TMdftb), intent(in) :: this
+
+    !> Number of dipole moment components
+    integer, intent(out) :: nDipole
+
+    !> Number of quadrupole moment components
+    integer, intent(out) :: nQuadrupole
+
+    integer, parameter :: dimDipole = 3
+    integer, parameter :: dimQuadrupole = 6
+
+    nDipole = dimDipole
+    nQuadrupole = dimQuadrupole
+
+  end subroutine getMultiExpanInfo
+
+
+  !> Returns the equivalence to get the correct mixing of charge dependent contributions
+  subroutine getOrbitalEquiv(this, equivDip, equivQuad)
+
+    !> Data structure
+    class(TMdftb), intent(inout) :: this
+
+    !> The equivalence vector for cumulative atomic dipole populations
+    integer, intent(out) :: equivDip(:,:)
+
+    !> The equivalence vector for cumulative atomic quadrupole populations
+    integer, intent(out) :: equivQuad(:,:)
+
+    integer :: nAtom, iCount, iSpin, nSpin, iAt, iSp, ii
+
+    nAtom = size(equivDip, dim=2)
+    nSpin = 1
+
+    equivDip(:,:) = 0
+    equivQuad(:,:) = 0
+
+    ! equivDip
+    iCount = 0
+    do iAt = 1, nAtom
+      do ii = 1, dimDipole
+        iCount = iCount + 1
+        equivDip(ii, iAt) = iCount
+      end do
+    end do
+
+    ! equivQuad
+    iCount = 0
+    do iAt = 1, nAtom
+      do ii = 1, dimQuadrupole
+        iCount = iCount + 1
+        equivQuad(ii, iAt) = iCount
+      end do
+    end do
```

Comment:
You have a lot of unused variables (`iSpin`, `nSpin`, `iSp`, etc.) Actually you could have written this in basically two lines IMO:
```
    equivDip(:,:) = reshape([(ii, ii = 1, dimDipol * nAtom)], [dimDipole, nAtom])
    equivQuad(:,:) = reshape([(ii, ii = 1, dimQuadrupole * nAtom)], [dimQuadrupole, nAtom])

```